File: packages/cordova-plugin-actito-geo/src/android/ActitoGeoPlugin.kt
package com.actito.geo.cordova

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.util.Log
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaArgs
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject
import com.actito.Actito
import com.actito.geo.ActitoGeo
import com.actito.geo.ktx.geo
import com.actito.geo.models.ActitoBeacon
import com.actito.geo.models.ActitoLocation
import com.actito.geo.models.ActitoRegion

class ActitoGeoPlugin : CordovaPlugin(), ActitoGeo.Listener {

    private companion object {
        private val TAG = ActitoGeoPlugin::class.java.simpleName
    }

    private var shouldShowRationale = false
    private var hasOnGoingPermissionRequest = false
    private var permissionRequestCallback: CallbackContext? = null

    private lateinit var permissionsLauncher: ActivityResultLauncher<Array<String>>

    override fun pluginInitialize() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        Actito.geo().addListener(this)

        permissionsLauncher = cordova.activity.registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions()
        ) { permissions ->
            val status = permissions
                .all { it.value }
                .let { granted ->
                    if (granted) {
                        PermissionStatus.GRANTED
                    } else {
                        if (!shouldShowRationale &&
                            !ActivityCompat.shouldShowRequestPermissionRationale(
                                cordova.activity,
                                permissions.keys.first()
                            )
                        ) {
                            PermissionStatus.PERMANENTLY_DENIED
                        } else {
                            PermissionStatus.DENIED
                        }
                    }
                }

            permissionRequestCallback?.success(status.rawValue)
            permissionRequestCallback = null
            hasOnGoingPermissionRequest = false
        }
    }

    override fun onDestroy() {
        Actito.geo().removeListener(this)
    }

    override fun execute(action: String, args: CordovaArgs, callback: CallbackContext): Boolean {
        when (action) {
            "hasLocationServicesEnabled" -> hasLocationServicesEnabled(args, callback)
            "hasBluetoothEnabled" -> hasBluetoothEnabled(args, callback)
            "getMonitoredRegions" -> getMonitoredRegions(args, callback)
            "getEnteredRegions" -> getEnteredRegions(args, callback)
            "enableLocationUpdates" -> enableLocationUpdates(args, callback)
            "disableLocationUpdates" -> disableLocationUpdates(args, callback)
            "checkPermissionStatus" -> checkPermissionStatus(args, callback)
            "shouldShowPermissionRationale" -> shouldShowPermissionRationale(args, callback)
            "presentPermissionRationale" -> presentPermissionRationale(args, callback)
            "requestPermission" -> requestPermission(args, callback)
            "openAppSettings" -> openAppSettings(args, callback)

            // Event broker
            "registerListener" -> registerListener(args, callback)

            else -> {
                callback.error("No implementation for action '$action'.")
                return false
            }
        }

        return true
    }

    // region Actito Geo

    private fun hasLocationServicesEnabled(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        callback.success(Actito.geo().hasLocationServicesEnabled)

    }

    private fun hasBluetoothEnabled(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        callback.success(Actito.geo().hasBluetoothEnabled)
    }

    private fun getMonitoredRegions(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        try {
            val json = JSONArray()
            Actito.geo().monitoredRegions.forEach { json.put(it.toJson()) }

            callback.success(json)
        } catch (e: Exception) {
            callback.error(e.message)
        }
    }

    private fun getEnteredRegions(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        try {
            val json = JSONArray()
            Actito.geo().enteredRegions.forEach { json.put(it.toJson()) }

            callback.success(json)
        } catch (e: Exception) {
            callback.error(e.message)
        }
    }

    private fun enableLocationUpdates(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Actito.geo().enableLocationUpdates()
        callback.void()
    }

    private fun disableLocationUpdates(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Actito.geo().disableLocationUpdates()
        callback.void()
    }

    private fun checkPermissionStatus(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val context = cordova.context ?: run {
            callback.error("Cannot continue without a context.")
            return
        }

        val permission =
            if (!args.isNull(0)) {
                val parameter = args.getString(0)
                PermissionGroup.parse(parameter) ?: run {
                    callback.error("Unsupported permission parameter: $parameter")
                    return
                }
            } else {
                callback.error("Missing permission parameter.")
                return
            }

        val status = determinePermissionStatus(context, permission)
        callback.success(status.rawValue)
    }

    private fun shouldShowPermissionRationale(
        @Suppress("UNUSED_PARAMETER") args: CordovaArgs,
        callback: CallbackContext
    ) {
        val activity = cordova.activity ?: run {
            logger.warning("Unable to acquire a reference to the current activity.")
            callback.error("Unable to acquire a reference to the current activity.")
            return
        }

        val permission =
            if (!args.isNull(0)) {
                val parameter = args.getString(0)
                PermissionGroup.parse(parameter) ?: run {
                    callback.error("Unsupported permission parameter: $parameter")
                    return
                }
            } else {
                callback.error("Missing permission parameter.")
                return
            }

        val manifestPermissions = getManifestPermissions(activity, permission)

        if (manifestPermissions.isEmpty()) {
            logger.warning("No permissions found in the manifest for $permission")
            callback.success(false)
            return
        }

        val showRationale = ActivityCompat.shouldShowRequestPermissionRationale(activity, manifestPermissions.first())
        callback.success(showRationale)
    }

    private fun presentPermissionRationale(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val activity = cordova.activity ?: run {
            logger.warning("Unable to acquire a reference to the current activity.")
            callback.error("Unable to acquire a reference to the current activity.")
            return
        }

        val permission =
            if (!args.isNull(0)) {
                val parameter = args.getString(0)
                PermissionGroup.parse(parameter) ?: run {
                    callback.error("Unsupported permission parameter: $parameter")
                    return
                }
            } else {
                callback.error("Missing permission parameter.")
                return
            }

        val rationale =
            if (!args.isNull(1)) {
                args.getJSONObject(1)
            } else {
                callback.error("Missing rationale parameter.")
                return
            }

        val title = if (!rationale.isNull("title")) rationale.getString("title") else null
        val message =
            if (!rationale.isNull("message")) {
                rationale.getString("message")
            } else {
                callback.error("Missing message parameter.")
                return
            }
        val buttonText =
            if (!rationale.isNull("buttonText")) rationale.getString("buttonText")
            else activity.getString(android.R.string.ok)

        try {
            logger.debug("Presenting permission rationale for '$permission'.")

            activity.runOnUiThread {
                AlertDialog.Builder(activity)
                    .setTitle(title)
                    .setMessage(message)
                    .setCancelable(false)
                    .setPositiveButton(buttonText, null)
                    .setOnDismissListener { callback.success() }
                    .show()
            }
        } catch (e: Exception) {
            callback.error("Unable to present the rationale alert.")
        }
    }

    private fun requestPermission(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val activity = cordova.activity ?: run {
            callback.error("Cannot continue without a activity.")
            return
        }

        val permission =
            if (!args.isNull(0)) {
                val parameter = args.getString(0)
                PermissionGroup.parse(parameter) ?: run {
                    callback.error("Unsupported permission parameter: $parameter")
                    return
                }
            } else {
                callback.error("Missing permission parameter.")
                return
            }

        if (hasOnGoingPermissionRequest) {
            logger.warning("A request for permissions is already running, please wait for it to finish before doing another request.")
            callback.error("A request for permissions is already running, please wait for it to finish before doing another request.")
            return
        }

        val status = determinePermissionStatus(activity, permission)
        if (status == PermissionStatus.GRANTED) {
            callback.success(status.rawValue)
            return
        }

        val manifestPermissions = getManifestPermissions(activity, permission)

        if (manifestPermissions.isEmpty()) {
            logger.warning("No permissions found in the manifest for $permission")
            callback.success(PermissionStatus.DENIED.rawValue)
            return
        }

        shouldShowRationale = ActivityCompat.shouldShowRequestPermissionRationale(activity, manifestPermissions.first())
        hasOnGoingPermissionRequest = true
        permissionRequestCallback = callback

        permissionsLauncher.launch(manifestPermissions.toTypedArray())
    }

    private fun openAppSettings(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        try {
            val context = cordova.context ?: run {
                callback.error("Cannot continue without a context.")
                return
            }

            val packageName = Uri.fromParts("package", context.packageName, null)
            context.startActivity(
                Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, packageName)
                    .addCategory(Intent.CATEGORY_DEFAULT)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    .addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY)
                    .addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS)
            )

            callback.void()
        } catch (e: Exception) {
            callback.error("Unable to open the app settings.")
        }
    }

    // endregion

    // region ActitoGeo.Listener

    override fun onLocationUpdated(location: ActitoLocation) {
        try {
            ActitoGeoPluginEventBroker.dispatchEvent("location_updated", location.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the location_updated event.", e)
        }
    }

    override fun onRegionEntered(region: ActitoRegion) {
        try {
            ActitoGeoPluginEventBroker.dispatchEvent("region_entered", region.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the region_entered event.", e)
        }
    }

    override fun onRegionExited(region: ActitoRegion) {
        try {
            ActitoGeoPluginEventBroker.dispatchEvent("region_exited", region.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the region_exited event.", e)
        }
    }

    override fun onBeaconEntered(beacon: ActitoBeacon) {
        try {
            ActitoGeoPluginEventBroker.dispatchEvent("beacon_entered", beacon.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the beacon_entered event.", e)
        }
    }

    override fun onBeaconExited(beacon: ActitoBeacon) {
        try {
            ActitoGeoPluginEventBroker.dispatchEvent("beacon_exited", beacon.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the beacon_exited event.", e)
        }
    }

    override fun onBeaconsRanged(region: ActitoRegion, beacons: List<ActitoBeacon>) {
        try {
            val payload = JSONObject()
            payload.put("region", region.toJson())
            payload.put("beacons", JSONArray().apply {
                beacons.forEach { put(it.toJson()) }
            })

            ActitoGeoPluginEventBroker.dispatchEvent("beacons_ranged", payload)
        } catch (e: Exception) {
            logger.error("Failed to emit the beacons_ranged event.", e)
        }
    }

    // endregion

    private fun registerListener(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        ActitoGeoPluginEventBroker.setup(preferences, object : ActitoGeoPluginEventBroker.Consumer {
            override fun onEvent(event: ActitoGeoPluginEventBroker.Event) {
                val payload = JSONObject()
                payload.put("name", event.name)
                when (event.payload) {
                    null -> {} // Skip encoding null payloads.
                    is Boolean -> payload.put("data", event.payload)
                    is Int -> payload.put("data", event.payload)
                    is Float -> payload.put("data", event.payload)
                    is Double -> payload.put("data", event.payload)
                    is String -> payload.put("data", event.payload)
                    is JSONObject -> payload.put("data", event.payload)
                    is JSONArray -> payload.put("data", event.payload)
                    else -> throw IllegalArgumentException("Unsupported event payload of type '${event.payload::class.java.simpleName}'.")
                }

                val result = PluginResult(PluginResult.Status.OK, payload)
                result.keepCallback = true

                callback.sendPluginResult(result)
            }
        })
    }

    private fun determinePermissionStatus(context: Context, permission: PermissionGroup): PermissionStatus {
        if (permission == PermissionGroup.BLUETOOTH) {
            return checkBluetoothPermissionStatus(context)
        }

        if (permission == PermissionGroup.BLUETOOTH_SCAN && Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            return checkBluetoothPermissionStatus(context)
        }

        val manifestPermissions = getManifestPermissions(context, permission)

        // If no permissions were found there's an issue and the permission is not set in the Android Manifest.
        if (manifestPermissions.isEmpty()) {
            logger.warning("No permissions found in the manifest for $permission")
            return PermissionStatus.DENIED
        }

        val granted = manifestPermissions.all {
            ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
        }

        return if (granted) PermissionStatus.GRANTED else PermissionStatus.DENIED
    }

    private fun getManifestPermissions(context: Context, permission: PermissionGroup): List<String> {
        val permissions = mutableListOf<String>()

        when (permission) {
            PermissionGroup.LOCATION_WHEN_IN_USE -> {
                if (hasPermissionInManifest(context, Manifest.permission.ACCESS_COARSE_LOCATION))
                    permissions.add(Manifest.permission.ACCESS_COARSE_LOCATION)

                if (hasPermissionInManifest(context, Manifest.permission.ACCESS_FINE_LOCATION))
                    permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
            }
            PermissionGroup.LOCATION_ALWAYS -> {
                // Note that the LOCATION_ALWAYS will deliberately request LOCATION_WHEN_IN_USE
                // case on pre Android Q devices. The ACCESS_BACKGROUND_LOCATION permission was only
                // introduced in Android Q, before it should be treated as the ACCESS_COARSE_LOCATION or
                // ACCESS_FINE_LOCATION.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    if (hasPermissionInManifest(context, Manifest.permission.ACCESS_BACKGROUND_LOCATION))
                        permissions.add(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
                } else {
                    permissions.addAll(getManifestPermissions(context, PermissionGroup.LOCATION_WHEN_IN_USE))
                }
            }
            PermissionGroup.BLUETOOTH_SCAN -> {
                // The BLUETOOTH_SCAN permission is introduced in Android S, meaning we should
                // not handle permissions on pre Android S devices.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    if (hasPermissionInManifest(context, Manifest.permission.BLUETOOTH_SCAN))
                        permissions.add(Manifest.permission.BLUETOOTH_SCAN)
                } else {
                    permissions.addAll(getManifestPermissions(context, PermissionGroup.BLUETOOTH))
                }
            }
            PermissionGroup.BLUETOOTH -> {
                if (hasPermissionInManifest(context, Manifest.permission.BLUETOOTH))
                    permissions.add(Manifest.permission.BLUETOOTH)
            }
        }

        return permissions
    }

    private fun hasPermissionInManifest(context: Context, permission: String): Boolean {
        try {
            val info = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                context.packageManager.getPackageInfo(
                    context.packageName,
                    PackageManager.PackageInfoFlags.of(PackageManager.GET_PERMISSIONS.toLong())
                )
            } else {
                context.packageManager.getPackageInfo(context.packageName, PackageManager.GET_PERMISSIONS)
            }
            if (info == null) {
                Log.d(TAG, "Unable to get Package info, will not be able to determine permissions to request.")
                return false
            }

            info.requestedPermissions?.let {
                for (r in it) {
                    if (r == permission) {
                        return true
                    }
                }
            }
        } catch (ex: Exception) {
            Log.d(TAG, "Unable to check manifest for permission: ", ex)
        }

        return false
    }

    private fun checkBluetoothPermissionStatus(context: Context): PermissionStatus {
        val manifestPermissions = getManifestPermissions(context, PermissionGroup.BLUETOOTH)

        if (manifestPermissions.isEmpty()) {
            logger.warning("Bluetooth permission missing in the manifest.")
            return PermissionStatus.DENIED
        }

        return PermissionStatus.GRANTED
    }

    internal enum class PermissionGroup {
        LOCATION_WHEN_IN_USE,
        LOCATION_ALWAYS,
        BLUETOOTH_SCAN,
        BLUETOOTH;

        internal val rawValue: String
            get() = when (this) {
                LOCATION_WHEN_IN_USE -> "location_when_in_use"
                LOCATION_ALWAYS -> "location_always"
                BLUETOOTH_SCAN -> "bluetooth_scan"
                BLUETOOTH -> "bluetooth"
            }

        internal companion object {
            internal fun parse(permission: String): PermissionGroup? {
                values().forEach {
                    if (it.rawValue == permission) return it
                }

                return null
            }
        }
    }

    internal enum class PermissionStatus {
        DENIED,
        GRANTED,
        PERMANENTLY_DENIED;

        internal val rawValue: String
            get() = when (this) {
                DENIED -> "denied"
                GRANTED -> "granted"
                PERMANENTLY_DENIED -> "permanently_denied"
            }

        internal companion object {
            internal fun parse(status: String): PermissionStatus? {
                values().forEach {
                    if (it.rawValue == status) return it
                }

                return null
            }
        }
    }
}

private fun CallbackContext.void() {
    sendPluginResult(PluginResult(PluginResult.Status.OK, null as String?))
}

private fun CallbackContext.success(b: Boolean) {
    sendPluginResult(PluginResult(PluginResult.Status.OK, b))
}

File: packages/cordova-plugin-actito-geo/src/ios/ActitoGeoPlugin.swift
import ActitoKit
import ActitoGeoKit
import CoreLocation

#if canImport(Cordova)
import Cordova
#endif

fileprivate let REQUESTED_LOCATION_ALWAYS_KEY = "com.actito.geo.cordova.requested_location_always"

@MainActor
@objc(ActitoGeoPlugin)
class ActitoGeoPlugin : CDVPlugin {
    private var locationManager: CLLocationManager!
    private var requestedPermission: PermissionGroup?
    private var requestedPermissionCall: CDVInvokedUrlCommand?

    private var authorizationStatus: CLAuthorizationStatus {
        if #available(iOS 14.0, *) {
            return locationManager.authorizationStatus
        }

        return CLLocationManager.authorizationStatus()
    }

    override func pluginInitialize() {
        super.pluginInitialize()

        loggerGeo.hasDebugLoggingEnabled = Actito.shared.options?.debugLoggingEnabled ?? false
        Actito.shared.geo().delegate = self

        locationManager = CLLocationManager()
        locationManager.delegate = self
    }

    @objc func registerListener(_ command: CDVInvokedUrlCommand) {
        let holdEventsUntilReady = self.commandDelegate.settings["com.actito.cordova.hold_events_until_ready"] as? String == "true"

        ActitoGeoPluginEventBroker.startListening(holdEventsUntilReady: holdEventsUntilReady) { event in
            var payload: [String: Any] = [
                "name": event.name,
            ]

            if let data = event.payload {
                payload["data"] = data
            }

            let result: CDVPluginResult = CDVPluginResult(status: .ok, messageAs: payload)
            result.keepCallback = true

            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    // MARK: - Actito Geo

    @objc func hasLocationServicesEnabled(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .ok, messageAs: Actito.shared.geo().hasLocationServicesEnabled)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func hasBluetoothEnabled(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .ok, messageAs: Actito.shared.geo().hasBluetoothEnabled)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func getMonitoredRegions(_ command: CDVInvokedUrlCommand) {
        do {
            let items = try Actito.shared.geo().monitoredRegions.map { try $0.toJson() }

            let result = CDVPluginResult(status: .ok, messageAs: items)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        } catch {
            let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    @objc func getEnteredRegions(_ command: CDVInvokedUrlCommand) {
        do {
            let items = try Actito.shared.geo().enteredRegions.map { try $0.toJson() }

            let result = CDVPluginResult(status: .ok, messageAs: items)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        } catch {
            let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    @objc func enableLocationUpdates(_ command: CDVInvokedUrlCommand) {
        Actito.shared.geo().enableLocationUpdates()

        let result = CDVPluginResult(status: .ok)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func disableLocationUpdates(_ command: CDVInvokedUrlCommand) {
        Actito.shared.geo().disableLocationUpdates()

        let result = CDVPluginResult(status: .ok)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func checkPermissionStatus(_ command: CDVInvokedUrlCommand) {
        guard let permissionStr = command.argument(at: 0) as? String, let permission = PermissionGroup(rawValue: permissionStr) else {
            let result = CDVPluginResult(status: .error, messageAs: "Missing 'permission' parameter.")
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        let status = checkPermissionStatus(permission)

        let result = CDVPluginResult(status: .ok, messageAs: status.rawValue)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func shouldShowPermissionRationale(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .ok, messageAs: false)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func presentPermissionRationale(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .error, messageAs: "This method is not implemented in iOS.")
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func requestPermission(_ command: CDVInvokedUrlCommand) {
        guard let permissionStr = command.argument(at: 0) as? String, let permission = PermissionGroup(rawValue: permissionStr) else {
            let result = CDVPluginResult(status: .error, messageAs: "Missing 'permission' parameter.")
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        let status = checkPermissionStatus(permission)

        if status == .granted || status == .permanentlyDenied || status == .restricted {
            let result = CDVPluginResult(status: .ok, messageAs: status.rawValue)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        if permission == .locationWhenInUse {
            if ((Bundle.main.infoDictionary?["NSLocationWhenInUseUsageDescription"]) == nil) {
                let result = CDVPluginResult(status: .error, messageAs: "Missing 'NSLocationWhenInUseUsageDescription' in the app bundle's Info.plist file.")
                self.commandDelegate!.send(result, callbackId: command.callbackId)
                return
            }
        }

        if permission == .locationAlways {
            if (Bundle.main.infoDictionary?["NSLocationAlwaysAndWhenInUseUsageDescription"]) == nil {
                let result = CDVPluginResult(status: .error, messageAs: "Missing 'NSLocationAlwaysAndWhenInUseUsageDescription' in the app bundle's Info.plist file.")
                self.commandDelegate!.send(result, callbackId: command.callbackId)
                return
            }

            if authorizationStatus == .notDetermined {
                let result = CDVPluginResult(status: .error, messageAs: "Location 'When in Use' should be requested before 'Location Always' request.")
                self.commandDelegate!.send(result, callbackId: command.callbackId)
                return
            }
        }

        requestedPermission = permission
        requestedPermissionCall = command

        if permission == .locationWhenInUse {
            locationManager.requestWhenInUseAuthorization()
        } else if permission == .locationAlways {
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(applicationDidBecomeActive),
                name: UIApplication.didBecomeActiveNotification,
                object: nil
            )

            locationManager.requestAlwaysAuthorization()
            UserDefaults.standard.set(true, forKey: REQUESTED_LOCATION_ALWAYS_KEY)
        }
    }

    @objc func openAppSettings(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString), UIApplication.shared.canOpenURL(url) else {
                let result = CDVPluginResult(status: .error, messageAs: "Unable to open the application settings.")
                self.commandDelegate!.send(result, callbackId: command.callbackId)
                return
            }

            UIApplication.shared.open(url) { success in
                if success {
                    let result = CDVPluginResult(status: .ok)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                } else {
                    let result = CDVPluginResult(status: .error, messageAs: "Unable to open the application settings.")
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            }
        }
    }

    // MARK: - Private API

    @objc private func applicationDidBecomeActive() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )

        guard let requestedPermissionCall = requestedPermissionCall else {
            return
        }

        if authorizationStatus != .authorizedAlways {
            let result = CDVPluginResult(status: .ok, messageAs: PermissionStatus.denied.rawValue)
            self.commandDelegate!.send(result, callbackId: requestedPermissionCall.callbackId)

            self.requestedPermission = nil
            self.requestedPermissionCall = nil
        }
    }

    private func checkPermissionStatus(_ permission: PermissionGroup) -> PermissionStatus {
        guard permission != .bluetoothScan else {
            return .granted
        }

        return determinePermissionStatus(permission, authorizationStatus: authorizationStatus)
    }

    private func determinePermissionStatus(_ permission: PermissionGroup, authorizationStatus: CLAuthorizationStatus) -> PermissionStatus {
        if permission == .locationAlways {
            switch authorizationStatus {
            case .notDetermined:
                return .denied
            case .restricted:
                return .restricted
            case .denied:
                return .permanentlyDenied
            case .authorizedWhenInUse:
                return UserDefaults.standard.bool(forKey: REQUESTED_LOCATION_ALWAYS_KEY) ? .permanentlyDenied : .denied
            case .authorizedAlways:
                return .granted
            @unknown default:
                return .denied
            }
        }

        switch authorizationStatus {
        case .notDetermined:
            return .denied
        case .restricted:
            return .restricted
        case .denied:
            return .permanentlyDenied
        case .authorizedWhenInUse, .authorizedAlways:
            return .granted
        @unknown default:
            return .denied
        }
    }
}

extension ActitoGeoPlugin: ActitoGeoDelegate {
    func actito(_ actitoGeo: ActitoGeo, didUpdateLocations locations: [ActitoLocation]) {
        guard let location = locations.first else { return }

        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "location_updated",
                payload: try location.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the location_updated event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didEnter region: ActitoRegion) {
        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "region_entered",
                payload: try region.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the region_entered event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didExit region: ActitoRegion) {
        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "region_exited",
                payload: try region.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the region_exited event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didEnter beacon: ActitoBeacon) {
        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "beacon_entered",
                payload: try beacon.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the beacon_entered event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didExit beacon: ActitoBeacon) {
        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "beacon_exited",
                payload: try beacon.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the beacon_exited event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didRange beacons: [ActitoBeacon], in region: ActitoRegion) {
        do {
            let payload: [String: Any] = [
                "region": try region.toJson(),
                "beacons": try beacons.map { try $0.toJson() },
            ]

            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "beacons_ranged",
                payload: payload
            )
        } catch {
            loggerGeo.error("Failed to emit the beacons_ranged event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didVisit visit: ActitoVisit) {
        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "visit",
                payload: try visit.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the visit event.", error: error)
        }
    }

    func actito(_ actitoGeo: ActitoGeo, didUpdateHeading heading: ActitoHeading) {
        do {
            ActitoGeoPluginEventBroker.dispatchEvent(
                name: "heading_updated",
                payload: try heading.toJson()
            )
        } catch {
            loggerGeo.error("Failed to emit the heading_updated event.", error: error)
        }
    }
}

extension ActitoGeoPlugin: CLLocationManagerDelegate {
    public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        handleAuthorizationChange(status)
    }

    @available(iOS 14.0, *)
    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        handleAuthorizationChange(manager.authorizationStatus)
    }

    private func handleAuthorizationChange(_ authorizationStatus: CLAuthorizationStatus) {
        if authorizationStatus == .notDetermined {
            // When the user changes to "Ask Next Time" via the Settings app.
            UserDefaults.standard.removeObject(forKey: REQUESTED_LOCATION_ALWAYS_KEY)
        }

        guard let requestedPermission = requestedPermission, let requestedPermissionCall = requestedPermissionCall else {
            return
        }

        let status = determinePermissionStatus(requestedPermission, authorizationStatus: authorizationStatus)

        // If status is 'Permanently Denied' we will consider that result as 'Denied' because the permission just changed
        // This will only affect 'When in Use' permission

        if status == .permanentlyDenied {
            let result = CDVPluginResult(status: .ok, messageAs: PermissionStatus.denied.rawValue)
            self.commandDelegate!.send(result, callbackId: requestedPermissionCall.callbackId)
        } else {
            let result = CDVPluginResult(status: .ok, messageAs: status.rawValue)
            self.commandDelegate!.send(result, callbackId: requestedPermissionCall.callbackId)
        }

        self.requestedPermission = nil
        self.requestedPermissionCall = nil
    }
}

extension ActitoGeoPlugin {
    internal enum PermissionGroup: String, CaseIterable {
        case locationWhenInUse = "location_when_in_use"
        case locationAlways = "location_always"
        case bluetoothScan = "bluetooth_scan"
    }

    internal enum PermissionStatus: String, CaseIterable {
        case denied = "denied"
        case granted = "granted"
        case restricted = "restricted"
        case permanentlyDenied = "permanently_denied"
    }
}
